### ISSUE TITLE
Can I make a reflective reference to an attribute slot?

### ISSUE THREAD

**Issue by vcschapp** [linkml/linkml#1469]
**Is your feature request related to a problem? Please describe.**

Granular data source tracking at the attribute (slot) level.

My group is building an open source data schema for an open data project. We will be publishing data objects which have data from various sources, introducing a need to track the data source of each data object. Moreover, we will be generating some of our data objects by **combining** data from **different sources**, leading to a need to track data sourcing at the sub-object (slot) level.

Assume we have two data objects like this:

```json
{
  "id": "object1",
  "source": "fooSource",
  "height": ["10", "m"]
}
```

and:

```json
{
  "id": "object2",
  "source": "barSource",
  "width": ["20", "m"]
}
```

We would like to be able to generate a composite data object with granular source tracking, along the lines of:

```json
{
  "id": "object12",
  "source": [
    {
      "property": "#/height",
      "source": "fooSource"
    },
    {
      "property": "#/width",
      "source": "barSource"
    }
  ],
  "height": ["10", "m"],
  "width": ["20", "m"]
}
```

Where just for the sake of argument `"#/height"` refers to the "height" slot of the combined object, and `#/width` refers to the "width" slot.

We would like the data validation process to validate that those references actually point somewhere legitimate. For example if the schema doesn't support a "height" slot, or if the schema has an optional "height" slot but the target data object doesn't actually have any data in that slot, then the validation would fail.

**Describe the solution you'd like**

- I would like the contents (value type) of a slot to be able to contain a reference to another slot.
- Think of the value stored in the slot as being like a LinkML-ified JSON Pointer or JSONPath reference.
- The LinkML data validator process would then validate that the slot value by ensuring that it is a valid reference expression which also points to a slot that exists in the schema.
- My use case is currently limited to "intra-class instance" references, *i.e.* a slot on a data instance of class A is referring to another slot of the same instance of class A.
    - It would need to be possible to refer to sub-slots at arbitrary levels of nesting, *i.e.* if an object which is an instance of class A has a slot "b" which contains an object which is an instance of class B, and class B has a slot named `foo` then I would like to be able to refer from the context of an instance of A to the slot `#/b/foo`, *i.e.* the `foo` slot in the instance of the "B" object which is stored in the `b` field of my instance of A.
    - One could imagine that there would also be inter-instance references, we just don't currently need them.

**How important is this feature?** Select from the options below:

• Important - it's a blocker and can't do work without it

Lack of this feature is close to being a launch blocker for us: it's an adoption blocker. My group is currently using JSON Schema and we are considering switching to LinkML. A switch to LinkML would be disruptive and cost resources, so we need to get a clear enough benefit to justify paying the cost to switch.

Having this feature really increases the value proposition for LinkML and may tip the balance, whereas missing it makes the adoption argument much more tenuous.

**When will use cases depending on this become relevant?** Select from the options below:

• Mid-term - 2-4 months
• Long-term - 6 months - 1 year

The timeframe for this becoming relevant to us is 4-8 months, as that is the realistic timeframe in which a switch away from JSON Schema to LinkML might be possible.

**Additional context**

**Comment by larryfreil** [comment]
I think an important use case to capture in the solution is the ability to reference to specific elements of an array combined from different sources. 

To extend the above examples, if we had additional data objects with address data for the object

```
{
  "id": "object3",
  "source": "fizzSource",
  "address": "120 Main St"
}
```
and
```
{
  "id": "object4",
  "source": "bazSource",
  "address": "100 Second St"
}
```
the result of all four would be something along the lines of:

```
{
  "id": "object1234",
  "source": [
    {
      "property": "#/height",
      "source": "fooSource"
    },
    {
      "property": "#/width",
      "source": "barSource"
    },
    {
      "property": "#/address[0]",
      "source": "fizzSource"
    },
    {
      "property": "#/address[1]",
      "source": "bazSource"
    }
  ],
  "height": ["10", "m"],
  "width": ["20", "m"],
  "address": ["120 Main St", "100 Second St"]
}
```

**Comment by vcschapp** [comment]
@larryfreil I agree. For the granular sourcing use case, the ability to look down into nested structures *including arrays* is key.

**Comment by nlharris** [comment]
I marked this "high priority" because of @vcschapp's comment
> Lack of this feature is close to being a launch blocker for us: it's an adoption blocker. 

However, this doesn't mean we necessarily have time to work on this feature soon. We are a small team and will do what we can!

**Comment by cmungall** [comment]
Thanks, this all makes sense.

I like the json-pointer based solution to this. LinkML has one foot in the linked data world, and the native RDF way to do this would be through reification. Unfortunately, reification is verbose; e.g it would look something like this:

```json
{
  "id": "object1234",
  "source": [
    {
      "property": "height",
      "source": "fooSource",
      "value": ["10", "m"],
    },
    {
      "property": "width",
      "source": "barSource"
      "value": ["20", "m"],
    },
    {
      "property": "address",
      "source": "fizzSource",
      "value": "120 Main St"
    },
    {
      "property": "address",
      "source": "bazSource"
      "value": "100 Second St"
    }
  ],
  "height": ["10", "m"],
  "width": ["20", "m"],
  "address": ["120 Main St", "100 Second St"]
}
```

combined with the lack of support for lists, this all gets fairly verbose, so I don't think we want to force people into an RDF solution here (but there should be a well-defined transform).

TODO: look at how this would be handled in https://json-ld.github.io/json-ld-star/

I think the pointer/path solution is elegant, I can see many use cases for which this kind of expressivity would be useful

**Comment by sierra-moxon** [comment]
It looks like the meta modeling components to support code that would eventually do a transformation like this, went into the last linkml release.  @vcschapp - have you had a chance to take a look at the proposal? :)

**Comment by vcschapp** [comment]
@sierra-moxon I have not yet, but will get this onto my agenda to review in the next day or two...

**Comment by vcschapp** [comment]
@sierra-moxon I took at the [changelog for v1.5.7](https://github.com/linkml/linkml/releases/tag/v1.5.7) look this morning but I think I'm looking in the wrong place.

- If there's a proposal to review, would you mind pointing me there?
- More for curiosity's sake, could you help me find the PR(s) that have the meta modeling components that would support our use case?

**Comment by nlharris** [comment]
Does anyone in the LinkML community have time to take a look at this?

**Comment by cmungall** [comment]
Current status:

- jsonpointer is a valid type in the linkml metamodel. Currently it just acts as a string, there is no additional validation
- Adding a validator should be easy. However, it's not yet clear if we should use an existing python jsonpointer implementation (these seem abandoned) or write our own?
- some semantics to clear up:
    - is there a use case for referring to a NULL? E.g. person1234 has no address, as determined by method9876
    - should json pointers be interpreted strictly over the json tree, or should non-lined references be traversed?

When we have decided these then we can unskip the validation behavior isn this test https://github.com/linkml/linkml/pull/1707
